core/packages/contracts/src/NativeTokens.sol

```diff
@@ -0,0 +1,149 @@
+
+    /* State */
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    bytes32 public immutable peerID;
+    bytes public peer;
+
+    TokenVault public immutable vault;
+    IOutboundChannel public outboundChannel;
+
+    uint256 public createTokenFee;
+
+    /* Errors */
+
+    error InvalidAmount();
+    error Unauthorized();
+    error NoFundsforCreateToken();
+
+    constructor(
+        TokenVault _vault,
+        IOutboundChannel _outboundChannel,
+        bytes memory _peer,
+        uint256 _createTokenFee
+    ) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        peer = _peer;
+        peerID = keccak256(_peer);
+        createTokenFee = _createTokenFee;
+    }
+
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message.
+    /// Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes calldata recipient, uint128 amount) external payable {
+        if (amount == 0) {
+            revert InvalidAmount();
+        }
+
+        vault.deposit(msg.sender, token, amount);
+
+        bytes memory payload = NativeTokensTypes.Mint(peer, token, recipient, amount);
+        outboundChannel.submit{ value: msg.value }(peer, payload);
+
+        emit Locked(recipient, token, amount);
+    }
+
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    function create(address token) external payable {
+        // to avoid spam, charge a fee for creating a new token
+        if (msg.value < createTokenFee) {
+            revert NoFundsforCreateToken();
+        }
+
+        IERC20Metadata metadata = IERC20Metadata(token);
+        bytes memory name = bytes(metadata.name());
+        if (name.length > 32) {
+            name = hex"";
+        }
+        bytes memory symbol = bytes(metadata.symbol());
+        if (symbol.length > 32) {
+            symbol = hex"";
+        }
+        uint8 decimals = metadata.decimals();
+
+        bytes memory payload = NativeTokensTypes.Create(peer, token, name, symbol, decimals);
+        outboundChannel.submit{ value: msg.value }(peer, payload);
+
+        emit Created(token);
+    }
+
+    /// @dev Processes messages from inbound channel.
+    /// @param origin The multilocation of the source parachain
+    /// @param message The message enqueued from substrate.
+    function handle(bytes calldata origin, bytes calldata message) external onlyRole(SENDER_ROLE) {
+        if (peerID != keccak256(origin)) {
+            revert Unauthorized();
+        }
```

Comment:
Should these 3 lines go into a separate, clearly defined function like "verifyPeerOrigin" ? Especially if we go with the above mentioned idea to create a reusable PeeredApp base class that other apps or other devs can use, then this function could go in that base class since it would likely be used across any peered app?